### PR TITLE
Export renderer in own namespace

### LINKED ISSUE #208: small improvements to renderers
- [ ] export renderers to be able to import them in applications
- [ ] make title (and other attributes) of renderers configurable (title is currently readonly most of the times) --> e.g. add getters and setters or pass via constructor or remove `readonly`

### PR DESCRIPTION
**Pull request by lehnerchristian**
closes #208

**prerequisites**: 
 * [x] branch is up-to-date with the branch to be merged with, i.e. develop
 * [ ] build is successful
 * [x] code is cleaned up and formatted 
 * [ ] tested with Firefox 52, Firefox 57+, Chrome 64+, MS Edge 16+

### Summary
*  Remove @internal marker from Renderer, which makes them publicly available in the .d.ts files

### CODE REVIEW

**Comment by lehnerchristian** [comment]
@sgratzl and @thinkh : I've removed the `@internal` markers, however the build is failing. I don't think that this is caused by this PR or at least I'm not sure.

I found a similar problem here https://github.com/Microsoft/TypeScript/issues/5711 and in this issue they state that we need more explicit type annotations

can one of you fix the type annotations if they are really causing our problem because I don't really know where to start at the moment

**Comment by thinkh** [comment]
I just noticed that LineUp v3 is using an older version of the lineupengine.

LineUp v3:
https://github.com/lineupjs/lineupjs/blob/85df597de875fc986e258df0c4d11fdfd5f72b24/package.json#L141

LineUp v4:
https://github.com/lineupjs/lineupjs/blob/0cfa2223b38980db5cd24c491d15a6b751938c07/package.json#L149

I don't know the changes between lineupengine v1 and v2. Hence, as @lehnerchristian suggested, we might need to change/add some typings over there.

**Comment by sgratzl** [comment]
> I just noticed that LineUp v3 is using an older version of the lineupengine.

on purpose since lineupengine v2 is designed to work with lineup v4

**Comment by thinkh** [comment]
@lehnerchristian Is this done from your side?

**Comment by lehnerchristian** [comment]
@thinkh , looks good as far as I can tell after this long time

**Comment by sgratzl** [comment]
what are the use cases again for exporting all renderers? In case of overriding them, most implement just the three factory functions, so in case one subclasses that, one doesn't gain much. Moreover, an alternative would be use just wrap one instance of an renderer with another one. Similar to the toolbar one can have access to all the registered renderers using the `defaultOptions`

e.g.

```js
const base = defaultOptions().renderers.number;

...
renderers: {
 'mynumber': {
    title: 'mynumber',
    canRender: base.canRender.bind(base,
   create: base.canRender.bind(base),
   createGroup: base.canRender.bind(base),
}

```

or 
```js
const base = defaultOptions().renderers.number;

class MyNumberR extends base.construtor {
...
}
```

**Comment by thinkh** [comment]
@sgratzl I understand your point that a renderer exposes only three methods and it might not be worth the opening. However, to me your suggested example looks like a different implementation pattern and unfamiliar if one is used to TypeScript where you can just import and extend a class or implement an interface. Furthermore, as a developer I would never guess that I can override or define a renderer like you suggested and would rather browse the API documentation.

@lehnerchristian and @puehringer Could you do the implementation in your app with the `defaultOptions()` that @sgratzl suggested? What are your opinions?

**Comment by puehringer** [comment]
@sgratzl @thinkh I indeed didn't know about these ways of renderer access/extensions. I will try to refactor the current implementation and check if it is more convenient this way. 
I still do not fully get why exporting them is not a more viable solution, as we also export the corresponding models too. I understand not exporting certain helper functions, but unique classes don't seem to be much of an issue?

**Comment by lehnerchristian** [comment]
I also didn't know / think of these ways. I'll try them too
however while this `const base = defaultOptions().renderers.number;` gives us an instance of the renderer we need and `base.constructor` gives us the class itself what's the difference to exporting the renderer class itself in the first place?

**Comment by sgratzl** [comment]
Models are complex objects with various methods and business logic. Renderers are simple and since it is a flat export of lineupjs it is already getting pretty big. Moreover, when I think about it, there is no need to have proper classes for renderers after all, since they are singleton instances and usually no inheritance. 

So, the more I export "officially" the less I can change in the future. like the call the base renderer in a composition style is easier than using it as a base class.

**Comment by sgratzl** [comment]
as a compromise one could expose them under a sub namespace. e.g. all the builder helper functions and classes are exported under the `.builderAdapter` namespace.

**Comment by thinkh** [comment]
@sgratzl Your compromise sounds like a plan. I'm not sure at the moment how to implement this. Can you provide an example of the namespace?

**Comment by sgratzl** [comment]
e.g. as in https://github.com/lineupjs/lineupjs/blob/master/src/builder/adapter/index.ts

**Comment by puehringer** [comment]
@sgratzl @thinkh I think the exporting in a certain namespace is a good idea, as I am currently facing some issues porting apps to lineup#v4. As already discussed, some functions were annotated with `@internal`, causes them to be not accessible. 

Some of these functions however are very essential (for advanced use-cases) such as: 
**Toolbar Utilities**
https://github.com/lineupjs/lineupjs/blob/lineup-v4/src/ui/toolbar.ts#L362-L373
https://github.com/lineupjs/lineupjs/blob/lineup-v4/src/ui/toolbar.ts#L376-L403
**CategoricalColumn**
https://github.com/lineupjs/lineupjs/blob/lineup-v4/src/model/internalCategorical.ts#L146-L160
...
These are just some functions which are used in apps and cannot be easily accessed/replicated. Implementing them on our own is not just a waste of resources, but also error prone and sometimes even impossible like [here](https://github.com/lineupjs/lineupjs/blob/lineup-v4/src/ui/toolbar.ts#L376-L403). 
I absolutely agree with you that we should not think _export first, regret later_, but certain functions such as `getToolbarDialogAddons` and `isCategoryIncluded` are essential and will most likely not change. 

Do you think it would be possible to (selectively) export functions in their particular namespace?

**Comment by sgratzl** [comment]
please open another issue or lets discuss that in person. I don't want to discuss things that are not PR related in a PR since the PR will be closed when it is merged.

**Comment by puehringer** [comment]
@sgratzl For the specific exports I will create separate tickets of course, but I was just listing a few examples supporting the namespace idea discussed above. 
Re: discussing things in person, what is your preferred way of contacting you? I think Github issues are good for summaries, but might not be the most efficient option.

**Comment by thinkh** [comment]
@sgratzl I reverted the `@internal` comments and exported the renderer classes and util functions in _./src/renderer/index.ts_. Please review if this is what you meant?

File: src/renderer/index.ts [hunk 1]
```diff
@@ -2,4 +2,109 @@ export {colorOf} from './impose';
 export * from './interfaces';
 export {renderMissingCanvas, renderMissingDOM} from './missing';
 
+import {AAggregatedGroupRenderer} from './AAggregatedGroupRenderer';
+import {ANumbersCellRenderer, matchRows} from './ANumbersCellRenderer';
+import {default as ActionRenderer} from './ActionRenderer';
```
Comment (sgratzl) [inline src/renderer/index.ts]: why not the regular `import ActionRenderer from './ActionRenderer';` syntax?

File: src/renderer/index.ts [hunk 2]
```diff
@@ -2,4 +2,109 @@ export {colorOf} from './impose';
...
+  SelectionRenderer,
+  SetCellRenderer,
+  SparklineCellRenderer,
+  StringCellRenderer,
+  TableCellRenderer,
+  UpSetCellRenderer,
+  VerticalBarCellRenderer
+};
+
+export const rendererUtils = {
```
Comment (sgratzl) [inline src/renderer/index.ts]: I don't see the point in exporting most of these functions cause they mainly exist for internal optimization and internal reuse. Knowing these functions takes more time than writing them by oneself. 

useful function imo:
 * adaptTextColorToBgColor
 * adaptDynamicColorToBgColor
 * toHeatMapColor

File: src/renderer/index.ts (hunk 1 above)
Comment (thinkh) [reply src/renderer/index.ts]: Makes sense. I changed it.

File: src/renderer/index.ts (hunk 2 above)
Comment (thinkh) [reply src/renderer/index.ts:80]: We discussed the exported functions again and reduced the list to some functions. Please check again.

**Comment by thinkh** [comment]
@sgratzl If you are fine with the changes, you can proceed and merge this PR.

**Pull request merged**